p2p/peersHolder/peersHolder.go

```diff
@@ -0,0 +1,199 @@
+	if len(knownSlice) == 0 {
+		return
+	}
+
+	pInfo := ph.connAddrSliceToPeerInfo[knownSlice]
+	if pInfo == nil {
+		ph.tempPeerIDsWaitingForShard[peerID] = knownSlice
+		ph.connAddrSliceToPeerInfo[knownSlice] = &peerInfo{
+			pid:     peerID,
+			shardID: 0, // this will be overwritten once shard is available
+		}
+
+		return
+	}
+
+	isOldData := peerID == pInfo.pid
+	if isOldData {
+		return
+	}
+
+	pInfo.pid = peerID
+}
+
+// PutShardID will perform the insert or the upgrade operation if the provided peerID is inside the preferred peers list
+func (ph *peersHolder) PutShardID(peerID core.PeerID, shardID uint32) {
+	ph.Lock()
+	defer ph.Unlock()
+
+	knownSlice, isWaitingForShardID := ph.tempPeerIDsWaitingForShard[peerID]
+	if !isWaitingForShardID {
+		return
+	}
+
+	pInfo, ok := ph.connAddrSliceToPeerInfo[knownSlice]
+	if !ok || pInfo == nil {
+		return
+	}
+
+	pInfo.shardID = shardID
+
+	ph.peerIDsPerShard[shardID] = append(ph.peerIDsPerShard[shardID], peerID)
+
+	ph.peerIDs[peerID] = &peerIDData{
+		connectionAddressSlice: knownSlice,
+		shardID:                shardID,
+		index:                  len(ph.peerIDsPerShard[shardID]) - 1,
+	}
+
+	delete(ph.tempPeerIDsWaitingForShard, peerID)
+}
+
+// Get will return a map containing the preferred peer IDs, split by shard ID
+func (ph *peersHolder) Get() map[uint32][]core.PeerID {
+	ph.RLock()
+	peerIDsPerShardCopy := ph.peerIDsPerShard
+	ph.RUnlock()
+
+	return peerIDsPerShardCopy
+}
+
+// Contains returns true if the provided peer id is a preferred connection
+func (ph *peersHolder) Contains(peerID core.PeerID) bool {
+	ph.RLock()
+	defer ph.RUnlock()
+
+	_, found := ph.peerIDs[peerID]
+	return found
+}
+
+// Remove will remove the provided peer ID from the inner members
+func (ph *peersHolder) Remove(peerID core.PeerID) {
+	ph.Lock()
+	defer ph.Unlock()
+
+	pidData, found := ph.peerIDs[peerID]
+	if !found {
+		return
+	}
+
+	shard, index, _ := ph.getShardAndIndexForPeer(peerID)
+	ph.removePeerFromMapAtIndex(shard, index)
+
+	connAddrSlice := pidData.connectionAddressSlice
+
+	delete(ph.peerIDs, peerID)
+
+	_, isPreferredPubKey := ph.connAddrSliceToPeerInfo[connAddrSlice]
+	if isPreferredPubKey {
+		// don't remove the entry because all the keys in this map refer to preferred connections and a reconnection might
+		// be done later
+		ph.connAddrSliceToPeerInfo[connAddrSlice] = nil
+	}
+
+	_, isWaitingForShardID := ph.tempPeerIDsWaitingForShard[peerID]
+	if isWaitingForShardID {
+		delete(ph.tempPeerIDsWaitingForShard, peerID)
+	}
+}
+
+// getKnownSlice checks if the connection address contains any of the initial preferred connection address slices
+// if true, it returns it
+// this function must be called under mutex protection
+func (ph *peersHolder) getKnownSlice(connectionAddressStr string) string {
+	for _, preferredConnAddrSlice := range ph.preferredConnAddrSlices {
+		if strings.Contains(connectionAddressStr, preferredConnAddrSlice) {
```

Comment:
I don't like these `strings.Contains`. Can we have something like a wrapper over the connection string types?
a component that handles a `/ip4/20.200.200.200/tcp/8080/p2p/some-random-pid`, a component that handles a  `/ip4/10.100.100.100`, plus a factory/builder that knows which instance to create.

Reply:
might leave a TODO for now, tho

Reply:
as discussed, the check on ctor would be enough